WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -373,12 +373,21 @@ def update
       return
     end
     registration_attributes = registration_params
+    was_accepted = @registration.accepted?
+    was_deleted = @registration.deleted?
+    # The only case we go to this endpoint if the registration was deleted is when we regiter again.
+    if was_deleted
+      # Set the accepted_at/deleted_at to nil iff it's not already set,
+      # which can happen when moving from deleted to accepted.
+      registration_attributes = { accepted_at: nil, deleted_at: nil }.merge(registration_attributes)
```

Comment:
Using two lines to do
```
registration_attributes[:accepted_at] = nil
registration_attributes[:deleted_at] = nil
```
is more verbose but also more explicit. I think it would be better for people who are not well-acquainted with Ruby.

Reply:
It would be more verbose, but unfortunately not semantically equivalent:
  - your suggestion says "take whatever is in `registration_attributes`, and override the two keys`.
  - mine says "take these two keys, and override with whatever is in `resgistration_attributes` (which is the behavior we want, given these keys may be set by `registration_params`).

Also I learnt ruby by looking at code that looked ruby-ish and not rubbish (mostly written by Jeremy, Jonatan, and al.), I'd argue that intentionally writing non ruby-ish code is actually a wrong bet for the readers; reading ruby-ish code they will certainly ask themselves what it does and hopefully learn some language feature from it.
Especially given that `Hash::merge` is a fairly common idiom among programming languages :)

Reply:
Okay... I accept your argument for Ruby-ish code from a perspective of motivating people to learn something new. But the fact that I screwed up trying to figure out what this line of code does kinda underlines my point that it is too packed/complex right now.

If there is no other concise way to solve this, feel free to discard my comment and resolve this conversation. But I just wanted to consider it in the first place.

Reply:
This is an alternative way of doing the same thing:
```suggestion
      unless registration_attributes.has_key?(:accepted_at)
        registration_attributes[:accepted_at] = nil
      end
      unless registration_attributes.has_key?(:deleted_at)
        registration_attributes[:deleted_at] = nil
      end
```

Now I'm generally used to receive comments like "your code does the thing, but did you know there is this construct in the language that does it for you?", but it's the first time I'm actually arguing to not change my code to something worse.

> But the fact that I screwed up trying to figure out what this line of code does kinda underlines my point that it is too packed/complex right now.

I don't agree with that:
  - there is a single method invocation on this line, which is definitely not "packed/complex" considering how often methods are chained in this repository.
  - it shows that until that PR the semantic of `a.merge(b)` was not clear for you.
It's worth noting that *specifically* to address this kind of situation and to make it clear to the reader, I wrote a comment that explain the intention of the line, and actually explain exactly what it does.

Since the suggestion I made at the beginning of this comment is semantically equivalent I'll let you pick whatever version you judge more suitable, my main concern is getting this merged asap since this bug is quite easy to trigger.

About the "write code readable for beginner point": if you really want this repository to follow this rule, I would suggest to make it clear somewhere what constructions we can or cannot use.
Indeed it's definitely hard to know where to draw the line between what construct is fine or not, and who draw that line.
For instance should we write this:
```ruby
new_array = []
for item in old_array do
  new_array << item.some_attribute
end
```
instead of `old_array.map(&:some_attribute)`, which is definitely not clear for someone who doesn't know ruby?

This map construct and the merge ones are quite used in our codebase (320 and 46 occurrences respectively), if we go this way and decide that some construct must be avoided, should we update all our codebase to get rid of them?

Reply:
Wow, this is a very thorough explanation. I appreciate and thank you for elaborating your argument, but at the same time I also apologise if I got you on the wrong foot there. It's absolutely fantastic that you stepped in to fix this bug and by no means did I intend to disregard your effort.

This PR will be merged after I finish writing this comment, but I want to make a point why I brought this up in the first place. This is not about a fundamental discussion which attributes/methods are "allowed" in our codebase and which ones aren't, and neither is it my attempt to become a "benevolent dictator" who decides what's readable enough and what isn't. It's simply my observation that in the current WST there are a lot of motivated members who keep struggling with Ruby and the style of code that Ruby deems idiomatic.

That being said, there is a significant difference between (trivial) `map` invocations and this `merge` invocation. Your example about `old_array.map(&:some_attribute)` reads almost like an (English) sentence would, in terms of "subject - verb - object". Take a bunch of elements and _map_ them to _some\_attribute_.
The merge construct doesn't follow that logic in that it first lists a hash with (multiple!) deliberate `nil` values and then packs a pre-filled hash on top of that. Unless you know the specifc implementation of `merge` in whatever programming language you happen to be using (so this isn't even my usual Ruby-bashing!), you cannot immediately derive the expected behavior when keys clash.
Those are all difficulties that do not arise with simpler idioms like `map` or `select`/`filter`, which is also reflected by the numbers you gave: The usages of `map` exceed `merge` by a factor of roughly 7 to 1.

Anyways, this is already too much of a debate about personal taste and I really don't want this detail discussion to taint the fact that I am very grateful for your quick and thorough bugfix :heart: Your argument has a strong point and you're getting my approval as a peace offering now :smile: